enhancements/substitutes-for.md

```diff
@@ -0,0 +1,379 @@
+A' substitutes for A
+A'' substitutes for A'
+
+We would first add the rule:
+A' -> A
+
+Since this is the first entry, there are no additional rules to generate.
+
+Next we add:
+A'' -> A'
+
+Which generates the following transitive substitution(D->C->B):
+A'' -> A
+
+So the total set of substitution rules is:
+A' -> A
+A'' -> A'
+A'' -> A
+
+Which means that ultimately B will be able to replace A, A', or A'', as desired.
+
+
+
+If we work the example in a different order we would first add the rule:
+A'' -> A
+
+Since this is the first entry, there are no additional rules to generate.
+
+Next we add:
+A' -> A
+
+Which generates the following transitive substitution (D->C->B):
+A'' -> A
+
+Giving us the same total set of rules:
+A' -> A
+A'' -> A'
+A'' -> A
+
+Note that the generator for the new rules is the result of a simple inner join query
+on the substitution table against itself, between the substitor and substitutee columns.
+
+Note also that each time a new rule is added, the table must be re-processed for additional
+new rules, until no new rules are generated.
+
+Alternatively the transitive substitution relationships could be derived at runtime when
+ListBundles is performed, following the substitution chains of any "skips" and "replaces"
+references to add the additional skips entries.
+
+Given the set of substitution relationships:
+A' -> A
+A'' -> A'
+A'' -> A
+
+As well as the replaces definition in the CSV (B replaces A), it remains for ListBundles
+to produce a CSV that will allow OLM to upgrade versions A'', A', and A' of the operator, 
+to version B.
+
+To achieve this, the CSV must be mutated as part of the ListBundles result.  Since
+"replaces" can only have a single entry, we will need to add additional "skips" entries
+to the CSV to account for the other valid upgrade paths.  Specifically, the mutated
+CSV for "B" will indicate that it skips A'' and A' (it already "replaces" A')
+
+In addition, the mutated version of A'' will add a skip for A.  (A'' will already skip A' as 
+provided by freshmaker).  Finally, A' will also skip A, as defined by freshmaker.
+
+This means the CSVs, as seen by OLM, will be:
+
+B: replaces A(original CSV), skips A', A''(result of substitution of A)
+A: whatever the original CSV listed
+A': skips A(added by freshmaker), plus original CSV
+A'': skips A'(added by freshmaker)+A(carried from CSV for A' when A' was rebuilt)+original CSV
+
+### User Stories [optional]
+
+Detail the things that people will be able to do if this is implemented.
+Include as much detail as possible so that people can understand the "how" of
+the system. The goal here is to make this feel real for users without getting
+bogged down.
+
+#### Freshmaker rebuilds a bundle that is the head of the channel, the bundle
+does not exist in any other channels.
+
+In this case, freshmaker simply takes the CSV it is rebuilding and adds a
+"skips" reference naming the bundle it is rebuilding.  A substitutes-for
```

Comment:
Why is the updated skips reference necessary? Doesn't opm take care of that because of the `substitutes-for` field?

Reply:
not sure what you mean by "opm" taking care of it, but if you mean that olm on the cluster should process the substitutes-for metadata and resolve things correctly, one of the goals here is that we do not have to patch olm on the cluster because that would then require that customers upgrade their clusters to an appropriate level for this to work.

So the CSV that is served to the cluster from the catalog must define the correct upgrade graph to the OLM resolver without depending on the value of "substitutes for"

Reply:
My point was that if we are already providing the `substitutes-for` field, then opm could infer that the bundle image being substituted should be skipped. If there is a good reason not to, then Freshmaker can make the adjustment instead.

Reply:
I think my reasoning for not doing that is that substituting-for something does not necessarily imply skipping it (or even necessarily imply any upgrade ordering relationship between the two)

Just because A + B are interchangeable (B substitutes for A) doesn't necessarily mean that B "skips" A.

So i would prefer that declaration be made explicitly, elsewhere.

Reply:
Okay, no problem.